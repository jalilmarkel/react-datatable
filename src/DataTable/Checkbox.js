import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { handleFunctionProps } from './util';

const baseCheckboxStyle = { fontSize: '18px', cursor: 'pointer' };

export default class Checkbox extends PureComponent {
  static propTypes = {
    indeterminate: PropTypes.bool,
    component: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.node,
      PropTypes.func,
    ]),
    onClick: PropTypes.func.isRequired,
    componentOptions: PropTypes.object,
    data: PropTypes.object,
    index: PropTypes.number,
    style: PropTypes.object,
  };

  static defaultProps = {
    indeterminate: false,
    component: 'input',
    componentOptions: {
      style: {},
    },
    data: {},
    index: null,
    style: null,
  };

  componentDidMount() {
    const { indeterminate } = this.props;

    this.el.indeterminate = indeterminate;
  }

  componentDidUpdate(prevProps) {
    const { indeterminate } = this.props;

    if (prevProps.indeterminate !== indeterminate) {
      this.el.indeterminate = indeterminate;
    }
  }

  handleClick = e => {
    const { onClick, data, index } = this.props;

    onClick(data, index, e);
  }

  render() {
    // remove indeterminate to prevent browser warnings
    const {
      component,
      componentOptions,
      indeterminate,
      ...rest
    } = this.props;

    const TagName = component || 'input';
    const baseStyle = TagName !== 'input' ? componentOptions.style : baseCheckboxStyle;
    const resolvedComponentOptions = handleFunctionProps(componentOptions, indeterminate);

    return (
      <TagName
        type="checkbox"
        {...rest}
        {...resolvedComponentOptions}
        // allow this component to fully control these options
        ref={el => { this.el = el; }}
        style={baseStyle}
        onClick={this.handleClick}
        onChange={() => null} // prevent uncontrolled checkbox warnings -  we don't need onChange
      />
    );
  }
}
