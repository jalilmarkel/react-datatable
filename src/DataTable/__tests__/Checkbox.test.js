import 'jest-styled-components';
import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import Checkbox from '../Checkbox';

afterEach(cleanup);

test('should render correctly when a custom component is not provided', () => {
  const { container } = render(<Checkbox name="test" onClick={jest.fn()} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox component/> should render correctly with a custom checkbox', () => {
  // eslint-disable-next-line react/prefer-stateless-function
  class CustomComp extends React.Component {
    render() {
      return (<div>25 schmeckles</div>);
    }
  }

  const { container } = render(<Checkbox name="test" onClick={jest.fn()} component={CustomComp} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox component/> should render correctly with custom props', () => {
  const { container } = render(<Checkbox name="test" onClick={jest.fn()} componentOptions={{ test: 'false' }} />);

  expect(container.firstChild).toMatchSnapshot();
});

test('component <Checkbox indeterminate /> should toggle indeterminate to true on the element', () => {
  const { container } = render(<Checkbox name="test" onClick={jest.fn()} indeterminate />);

  expect(container.firstChild.indeterminate).toBe(true);
});

test('component <Checkbox indeterminate={false} /> should not toggle indeterminate if theere is no change', () => {
  const { container } = render(<Checkbox name="test" onClick={jest.fn()} indeterminate={false} />);

  expect(container.firstChild.indeterminate).toBe(false);
});

test('should handle onClick', () => {
  const mockCallback = jest.fn();
  const { container } = render(<Checkbox name="test" data={{ name: 'morty' }} onClick={mockCallback} />);
  fireEvent.click(container.firstChild);

  expect(mockCallback).toBeCalled();
  expect(mockCallback.mock.calls[0][0]).toEqual({ name: 'morty' });
  expect(mockCallback.mock.calls[0][1]).toBeDefined(); // TODO: mock event?
});
