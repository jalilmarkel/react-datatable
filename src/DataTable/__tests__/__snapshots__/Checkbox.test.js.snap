// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <Checkbox /> should render correctly 1`] = `
<Checkbox
  component="input"
  componentOptions={
    Object {
      "style": Object {},
    }
  }
  data={Object {}}
  indeterminate={false}
  index={null}
  onClick={[MockFunction]}
  style={null}
>
  <input
    data={Object {}}
    index={null}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontSize": "18px",
      }
    }
    type="checkbox"
  />
</Checkbox>
`;

exports[`component <Checkbox /> should render correctly when a custom component is not provided 1`] = `
<Checkbox
  component="input"
  componentOptions={
    Object {
      "style": Object {},
    }
  }
  data={Object {}}
  indeterminate={false}
  index={null}
  onClick={[MockFunction]}
  style={null}
>
  <input
    data={Object {}}
    index={null}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontSize": "18px",
      }
    }
    type="checkbox"
  />
</Checkbox>
`;

exports[`component <Checkbox component/> should render correctly with a custom checkbox 1`] = `
<Checkbox
  component={[Function]}
  componentOptions={
    Object {
      "style": Object {},
    }
  }
  data={Object {}}
  indeterminate={false}
  index={null}
  onClick={[MockFunction]}
  style={null}
>
  <CustomComp
    data={Object {}}
    index={null}
    onChange={[Function]}
    onClick={[Function]}
    style={Object {}}
    type="checkbox"
  >
    <div>
      25 schmeckles
    </div>
  </CustomComp>
</Checkbox>
`;

exports[`component <Checkbox component/> should render correctly with custom props 1`] = `
<Checkbox
  component="input"
  componentOptions={
    Object {
      "test": "false",
    }
  }
  data={Object {}}
  indeterminate={false}
  index={null}
  onClick={[MockFunction]}
  style={null}
>
  <input
    data={Object {}}
    index={null}
    onChange={[Function]}
    onClick={[Function]}
    style={
      Object {
        "cursor": "pointer",
        "fontSize": "18px",
      }
    }
    test="false"
    type="checkbox"
  />
</Checkbox>
`;
